src/backend/distributed/planner/recursive_planning.c

```diff
@@ -599,6 +616,303 @@ RecursivelyPlanNonColocatedSubqueriesInWhere(Query *query,
+				if (leftNodeRecurs && !rightNodeRecurs)
+				{
+					RecursivelyPlanDistributedJoinNode(rightNode, query,
+													   recursivePlanningContext);
+					return;
+				}
+
+				break;
+			}
+
+			case JOIN_RIGHT:
+			{
+				bool leftNodeRecurs = IsJoinNodeRecurring(leftNode, query);
+				bool rightNodeRecurs = IsJoinNodeRecurring(rightNode, query);
+
+				/* <distributed> right join <recurring> */
+				if (!leftNodeRecurs && rightNodeRecurs)
+				{
+					RecursivelyPlanDistributedJoinNode(leftNode, query,
+													   recursivePlanningContext);
+					return;
+				}
+
+				break;
+			}
+
+			case JOIN_FULL:
+			{
+				bool leftNodeRecurs = IsJoinNodeRecurring(leftNode, query);
+				bool rightNodeRecurs = IsJoinNodeRecurring(rightNode, query);
+
+				/*
+				 * <recurring> full join <distributed>
+				 * <distributed> full join <recurring>
+				 */
+				if (leftNodeRecurs && !rightNodeRecurs)
+				{
+					RecursivelyPlanDistributedJoinNode(rightNode, query,
+													   recursivePlanningContext);
+					return;
+				}
+				else if (!leftNodeRecurs && rightNodeRecurs)
+				{
+					RecursivelyPlanDistributedJoinNode(leftNode, query,
+													   recursivePlanningContext);
+					return;
+				}
+
+				break;
+			}
+
+			case JOIN_INNER:
+			{
+				/*
+				 * We don't need to recursively plan non-outer joins and we already
+				 * descended into sub join trees to handle outer joins buried in them.
+				 */
+				break;
+			}
+
+			default:
+			{
+				ereport(ERROR, (errmsg("got unexpected join type (%d) when recursively "
+									   "planning a join",
+									   joinExpr->jointype)));
+			}
+		}
+	}
+	else
+	{
+		ereport(ERROR, errmsg("got unexpected expr type when recursively "
+							  "planning a join"));
+	}
+}
+
+
+/*
+ * IsJoinNodeRecurring determines whether a join node in a join tree
+ * is recurring.
+ */
+static bool
+IsJoinNodeRecurring(Node *joinNode, Query *query)
+{
+	if (IsA(joinNode, JoinExpr))
+	{
+		/*
+		 * Then this is a sub (join) tree and here is how we determine whether a
+		 * join tree is recurring or not:
+		 *
+		 * i)   Outer join:
+		 *
+		 *      If the outer side of the join is a recurring rel, then the join
+		 *      node itself is recurring too regardless of whether the inner
+		 *      side of the join is recurring or not.
+		 *
+		 *      If outer side of the join is distributed, then the join
+		 *      node is distributed too.
+		 *
+		 * ii)  Inner join:
+		 *
+		 *      The join node is recurring only if both sides of the join are
+		 *      recurring.
+		 *
+		 *      Otherwise, if at least one side of the join is distributed, then
+		 *      this would ensure not returning recurring tuples from the workers;
+		 *      hence makes the join node "not recurring".
+		 */
+		JoinExpr *joinExpr = (JoinExpr *) joinNode;
+		bool leftNodeIsRecurring = IsJoinNodeRecurring(joinExpr->larg, query);
+		bool rightNodeIsRecurring = IsJoinNodeRecurring(joinExpr->rarg, query);
+
+		switch (joinExpr->jointype)
```

Comment:
can we avoid IsJoinNodeRecurring when possible? Say for RIGHT_JOIN do not check leftNodeIsRecurring etc.

Reply:
If we consider the example scenario: left_rel RIGTH JOIN right_rel;

Here we need to recursively plan left_rel only if it's distributed and the other side is recurring; meaning that we need to check both sides of the join tree to decide.

Reply:
But what we can do is that we can drop `bool leftNodeRecurs = ` variables and instead directly call `IsJoinNodeRecurring()` functions in `if` statements so that --in the example scenario-- we don't need to check the right node if the left one is recurring.

Reply:
77a2842 fixes this too.